### PR TITLE
feat(dns): DNS support multi line record set resource

### PR DESCRIPTION
**Pull request by deer-hang**
**What this PR does / why we need it**:
 DNS support multi line record set resource
**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #xxx

**Special notes for your reviewer**:

**Release note**:

```release-note

```

## PR Checklist

* [X] Tests added/passed.
* [X] Documentation updated.
* [X] Schema updated.

## Acceptance Steps Performed

* Change exist dns recordset resource to V2
```
make testacc TEST='./huaweicloud/services/acceptance/dns' TESTARGS='-run TestAccDNSV2RecordSet_'
==> Checking that code complies with gofmt requirements... 
TF_ACC=1 go test ./huaweicloud/services/acceptance/dns -v -run TestAccDNSV2RecordSet_ -timeout 360m -parallel 4 
=== RUN   TestAccDNSV2RecordSet_basic 
=== PAUSE TestAccDNSV2RecordSet_basic 
=== RUN   TestAccDNSV2RecordSet_readTTL
=== PAUSE TestAccDNSV2RecordSet_readTTL
=== RUN   TestAccDNSV2RecordSet_private
=== PAUSE TestAccDNSV2RecordSet_private
=== CONT  TestAccDNSV2RecordSet_basic
=== CONT  TestAccDNSV2RecordSet_private
=== CONT  TestAccDNSV2RecordSet_readTTL
--- PASS: TestAccDNSV2RecordSet_private (37.46s) 
--- PASS: TestAccDNSV2RecordSet_readTTL (37.85s) 
--- PASS: TestAccDNSV2RecordSet_basic (63.57s) 
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/dns       63.634s
```

* Add new dns recordset resource
```
make testacc TEST='./huaweicloud/services/acceptance/dns' TESTARGS='-run TestAccDNSRecordset_basic'
==> Checking that code complies with gofmt requirements... 
TF_ACC=1 go test ./huaweicloud/services/acceptance/dns -v -run TestAccDNSRecordset_basic -timeout 360m -parallel 4 
=== RUN   TestAccDNSRecordset_basic 
=== PAUSE TestAccDNSRecordset_basic
=== CONT  TestAccDNSRecordset_basic
--- PASS: TestAccDNSRecordset_basic (50.69s) 
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/dns       50.748s
```

### CODE REVIEW

File: huaweicloud/services/acceptance/dns/resource_huaweicloud_dns_recordset_test.go [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+resource "huaweicloud_dns_recordset" "test" {
+  zone_id = huaweicloud_dns_zone.zone_1.id
+  name = "%s"
+  type = "A"
+  description = "a recordset description"
+  status = "ENABLE"
+  ttl = 300
+  records = ["10.1.0.0"]
+  line_id = "Dianxin_Shanxi"
+  weight = 3
```
Comment (houpeng80) [inline huaweicloud/services/acceptance/dns/resource_huaweicloud_dns_recordset_test.go]: Modify the format to ensure equal sign alignment

File: huaweicloud/services/acceptance/dns/resource_huaweicloud_dns_recordset_test.go [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+resource "huaweicloud_dns_recordset" "test" {
+  zone_id = huaweicloud_dns_zone.zone_1.id
+  name = "update.%s"
+  type = "TXT"
+  description = "a recordset description update"
+  status = "DISABLE"
+  ttl = 600
+  records = ["\"test records\""]
+  line_id = "Dianxin_Shanxi"
+  weight = 5
```
Comment (houpeng80) [inline huaweicloud/services/acceptance/dns/resource_huaweicloud_dns_recordset_test.go]: ditto

File: huaweicloud/services/dns/resource_huaweicloud_dns_recordset.go [hunk 3]
```diff
@@ -0,0 +1,482 @@
...
+				ValidateFunc: validation.StringInSlice([]string{
+					"A", "AAAA", "MX", "CNAME", "TXT", "NS", "SRV", "CAA",
+				}, false),
+				Description: `Specifies the type of the record set.`,
+			},
+			"records": {
+				Type:        schema.TypeList,
+				Elem:        &schema.Schema{Type: schema.TypeString},
+				MinItems:    1,
+				Required:    true,
```
Comment (houpeng80) [inline huaweicloud/services/dns/resource_huaweicloud_dns_recordset.go:76]: this param is not required in the API

File: huaweicloud/services/dns/resource_huaweicloud_dns_recordset.go [hunk 4]
```diff
@@ -0,0 +1,482 @@
...
+				Optional:     true,
+				Computed:     true,
+				ValidateFunc: validation.IntBetween(0, 1000),
+				Description:  `Specifies the weight of the record set.`,
+			},
+			"tags": common.TagsSchema(),
+			"zone_name": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: `The zone name of the record set.`,
```
Comment (houpeng80) [inline huaweicloud/services/dns/resource_huaweicloud_dns_recordset.go:110]: the description content should start with Specifies

File: docs/resources/dns_recordset.md [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+-> **NOTE:** You can use custom line or get more information about default resolution lines
+from [Resolution Lines](https://support.huaweicloud.com/intl/en-us/api-dns/en-us_topic_0085546214.html).
+
+* `name` - (Required, String) Specifies the name of the record set.
+  The name suffixed with a zone name, which is a complete host name ended with a dot.
+
+* `type` - (Required, String) Specifies the type of the record set.
+  Value options: **A**, **AAAA**, **MX**, **CNAME**, **TXT**, **NS**, **SRV**, **CAA**.
+
+* `records` - (Required, List) Specifies an array of DNS records. The value rules vary depending on the record set type.
```
Comment (houpeng80) [inline docs/resources/dns_recordset.md:60]: this param is not required in the API

File: huaweicloud/services/dns/resource_huaweicloud_dns_recordset.go [hunk 6]
```diff
@@ -0,0 +1,482 @@
...
+
+	// set tags
+	if err := utils.SetResourceTagsToState(d, getDNSRecordsetClient, "DNS-public_recordset"); err != nil {
+		return diag.FromErr(err)
+	}
+	return nil
+}
+
+func getDNSRecordsetStatus(getDNSRecordsetRespBody interface{}) string {
+	status := utils.PathSearch("status", getDNSRecordsetRespBody, nil).(string)
```
Comment (houpeng80) [inline huaweicloud/services/dns/resource_huaweicloud_dns_recordset.go]: status := utils.PathSearch("status", getDNSRecordsetRespBody, nil).(string)
status := utils.PathSearch("status", getDNSRecordsetRespBody, "").(string)

File: huaweicloud/services/dns/resource_huaweicloud_dns_recordset.go [hunk 7]
```diff
@@ -0,0 +1,482 @@
...
+		Target:       []string{"DELETED"},
+		Pending:      []string{"ACTIVE", "PENDING", "ERROR"},
+		Refresh:      dnsRecordsetStatusRefreshFunc(d, recordsetClient),
+		Timeout:      10 * time.Minute,
+		Delay:        5 * time.Second,
+		PollInterval: 5 * time.Second,
+	}
+	_, err := stateConf.WaitForStateContext(ctx)
+	if err != nil {
+		return fmt.Errorf("error waiting for DNS recordset (%s) to become DELETED: %s", d.Id(), err)
```
Comment (houpeng80) [inline huaweicloud/services/dns/resource_huaweicloud_dns_recordset.go]: It is more appropriate to modify "become" to "be"

File: huaweicloud/services/dns/resource_huaweicloud_dns_recordset.go [hunk 8]
```diff
@@ -0,0 +1,482 @@
...
+	return nil
+}
+
+func waitForDNSRecordsetCreateOrUpdate(ctx context.Context, recordsetClient *golangsdk.ServiceClient,
+	d *schema.ResourceData) error {
+	stateConf := &resource.StateChangeConf{
+		Target:       []string{"ACTIVE", "DISABLE"},
+		Pending:      []string{"PENDING"},
+		Refresh:      dnsRecordsetStatusRefreshFunc(d, recordsetClient),
+		Timeout:      10 * time.Minute,
```
Comment (houpeng80) [inline huaweicloud/services/dns/resource_huaweicloud_dns_recordset.go]: the timeout can be declared uniformly in the start of the file

File: huaweicloud/services/dns/resource_huaweicloud_dns_recordset.go [hunk 9]
```diff
@@ -0,0 +1,482 @@
...
+	return nil
+}
+
+func waitForDNSRecordsetDeleted(ctx context.Context, recordsetClient *golangsdk.ServiceClient,
+	d *schema.ResourceData) error {
+	stateConf := &resource.StateChangeConf{
+		Target:       []string{"DELETED"},
+		Pending:      []string{"ACTIVE", "PENDING", "ERROR"},
+		Refresh:      dnsRecordsetStatusRefreshFunc(d, recordsetClient),
+		Timeout:      10 * time.Minute,
```
Comment (houpeng80) [inline huaweicloud/services/dns/resource_huaweicloud_dns_recordset.go]: ditto

File: huaweicloud/services/dns/resource_huaweicloud_dns_recordset.go (hunk 6 above)
Comment (deer-hang) [reply huaweicloud/services/dns/resource_huaweicloud_dns_recordset.go]: OK

File: huaweicloud/services/dns/resource_huaweicloud_dns_recordset.go (hunk 7 above)
Comment (deer-hang) [reply huaweicloud/services/dns/resource_huaweicloud_dns_recordset.go]: OK

File: huaweicloud/services/dns/resource_huaweicloud_dns_recordset.go (hunk 8 above)
Comment (deer-hang) [reply huaweicloud/services/dns/resource_huaweicloud_dns_recordset.go]: OK

File: huaweicloud/services/dns/resource_huaweicloud_dns_recordset.go (hunk 9 above)
Comment (deer-hang) [reply huaweicloud/services/dns/resource_huaweicloud_dns_recordset.go]: OK

**Comment by deer-hang** [comment]
```
make testacc TEST='./huaweicloud/services/acceptance/dns' TESTARGS='-run TestAccDNSRecordset_basic'
==> Checking that code complies with gofmt requirements... 
TF_ACC=1 go test ./huaweicloud/services/acceptance/dns -v -run TestAccDNSRecordset_basic -timeout 360m -parallel 4 
=== RUN   TestAccDNSRecordset_basic 
=== PAUSE TestAccDNSRecordset_basic
=== CONT  TestAccDNSRecordset_basic
--- PASS: TestAccDNSRecordset_basic (49.92s) 
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/dns       49.970s
```

File: huaweicloud/services/dns/resource_huaweicloud_dns_recordset.go [hunk 10]
```diff
@@ -0,0 +1,490 @@
...
+				Computed: true,
+				ForceNew: true,
+			},
+			"zone_id": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: `Specifies the public zone ID.`,
+			},
+			"line_id": {
```
Comment (ShiChangkuo) [inline huaweicloud/services/dns/resource_huaweicloud_dns_recordset.go]: please sort parameters in sort of Required, Optional and Computed

File: huaweicloud/services/dns/resource_huaweicloud_dns_recordset.go [hunk 11]
```diff
@@ -0,0 +1,490 @@
...
+	createDNSRecordsetRespBody, err := utils.FlattenResponse(createDNSRecordsetResp)
+	if err != nil {
+		return err
+	}
+
+	id, err := jmespath.Search("id", createDNSRecordsetRespBody)
+	if err != nil {
+		return fmt.Errorf("error creating DNS recordset: ID is not found in API response")
+	}
+	d.SetId(id.(string))
```
Comment (ShiChangkuo) [inline huaweicloud/services/dns/resource_huaweicloud_dns_recordset.go]: it's better to set the ID in format of `zone_id/id` to keep compatible with v2 resource.

File: docs/resources/dns_recordset.md [hunk 12]
```diff
@@ -4,12 +4,10 @@ subcategory: "Domain Name Service (DNS)"
 
 # huaweicloud_dns_recordset
 
-Manages a DNS record set in the HuaweiCloud DNS Service.
+Manages a DNS record set resource within HuaweiCloud.
 
 ## Example Usage
 
-### Automatically detect the correct network
```
Comment (ShiChangkuo) [inline docs/resources/dns_recordset.md:11]: please add another example for multi-line

**Comment by deer-hang** [comment]
```
make testacc TEST='./huaweicloud/services/acceptance/dns' TESTARGS='-run TestAccDNSRecordset_'
==> Checking that code complies with gofmt requirements... 
TF_ACC=1 go test ./huaweicloud/services/acceptance/dns -v -run TestAccDNSRecordset_ -timeout 360m -parallel 4 
=== RUN   TestAccDNSRecordset_basic 
=== PAUSE TestAccDNSRecordset_basic
=== RUN   TestAccDNSRecordset_publicZone
=== PAUSE TestAccDNSRecordset_publicZone
=== RUN   TestAccDNSRecordset_privateZone
=== PAUSE TestAccDNSRecordset_privateZone
=== CONT  TestAccDNSRecordset_basic
=== CONT  TestAccDNSRecordset_privateZone
=== CONT  TestAccDNSRecordset_publicZone
--- PASS: TestAccDNSRecordset_publicZone (48.81s) 
--- PASS: TestAccDNSRecordset_basic (50.26s) 
--- PASS: TestAccDNSRecordset_privateZone (52.97s) 
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/dns       53.025s
```

File: docs/resources/dns_recordset.md [hunk 13]
```diff
@@ -4,11 +4,11 @@ subcategory: "Domain Name Service (DNS)"
...
+### Record Set Multi-line
```
Comment (ShiChangkuo) [inline docs/resources/dns_recordset.md]: Perhaps `Multi-line Record Set` or `Record Set with Multi-line` is more appropriate.

File: docs/resources/dns_recordset.md [hunk 14]
```diff
@@ -33,45 +83,59 @@ resource "huaweicloud_dns_recordset" "rs_example_com" {
...
-* `description` - (Optional, String) A description of the record set.
+* `status` - (Optional, String) Specifies the status of the record set.
+  Value options: **ENABLE**, **DISABLE**. The default value is **ENABLE**.
 
-* `tags` - (Optional, Map) The key/value pairs to associate with the record set.
+* `tags` - (Optional, Map) Specifies the key/value pairs to associate with the DNS recordset.
+
+* `description` - (Optional, String) Specifies the description of the record set.
+
+* `weight` - (Optional, Int) Specifies the weight of the record set.
```
Comment (ShiChangkuo) [inline docs/resources/dns_recordset.md:116]: it's better to put `line_id` and `weight` together.

File: huaweicloud/services/dns/resource_huaweicloud_dns_recordset.go [hunk 15]
```diff
@@ -0,0 +1,592 @@
...
+	if len(idArrays) != 2 {
+		err = fmt.Errorf("invalid format specified for ID. Format must be <zone_id>/<recordset_id>")
+		return
+	}
+	zoneID = idArrays[0]
+	recordsetID = idArrays[1]
+	return
+}
+
+func getVersionByZoneType(zoneType string) string {
```
Comment (ShiChangkuo) [inline huaweicloud/services/dns/resource_huaweicloud_dns_recordset.go]:
```suggestion
func getApiVersionByZoneType(zoneType string) string {
```

File: huaweicloud/services/dns/resource_huaweicloud_dns_recordset.go [hunk 16]
```diff
@@ -0,0 +1,592 @@
...
+	}
+	return nil
+}
+
+func setDNSRecordsetTags(d *schema.ResourceData, client *golangsdk.ServiceClient, id, zoneType string) error {
+	resourceType, err := utils.GetDNSRecordSetTagType(zoneType)
+	if err != nil {
+		return err
+	}
+	if resourceTags, err := tags.Get(client, resourceType, id).Extract(); err == nil {
```
Comment (ShiChangkuo) [inline huaweicloud/services/dns/resource_huaweicloud_dns_recordset.go]: we can use utils.SetResourceTagesToState instead of this block

**Comment by deer-hang** [comment]
```
make testacc TEST='./huaweicloud/services/acceptance/dns' TESTARGS='-run TestAccDNSRecordset_'
==> Checking that code complies with gofmt requirements... 
TF_ACC=1 go test ./huaweicloud/services/acceptance/dns -v -run TestAccDNSRecordset_ -timeout 360m -parallel 4 
=== RUN   TestAccDNSRecordset_basic 
=== PAUSE TestAccDNSRecordset_basic
=== RUN   TestAccDNSRecordset_publicZone
=== PAUSE TestAccDNSRecordset_publicZone
=== RUN   TestAccDNSRecordset_privateZone
=== PAUSE TestAccDNSRecordset_privateZone
=== CONT  TestAccDNSRecordset_basic
=== CONT  TestAccDNSRecordset_privateZone
=== CONT  TestAccDNSRecordset_publicZone
--- PASS: TestAccDNSRecordset_publicZone (52.74s) 
--- PASS: TestAccDNSRecordset_basic (54.42s) 
--- PASS: TestAccDNSRecordset_privateZone (57.75s) 
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/dns       57.805s
```

**Pull request merged**